File: src/main/java/me/ruslanys/vkmusic/services/impl/DefaultDownloadService.java
package me.ruslanys.vkmusic.services.impl;

import lombok.NonNull;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import me.ruslanys.vkmusic.component.VkClient;
import me.ruslanys.vkmusic.component.impl.DownloadTask;
import me.ruslanys.vkmusic.entity.Audio;
import me.ruslanys.vkmusic.entity.domain.DownloadResult;
import me.ruslanys.vkmusic.entity.domain.event.DownloadFailEvent;
import me.ruslanys.vkmusic.entity.domain.event.DownloadFinishEvent;
import me.ruslanys.vkmusic.entity.domain.event.DownloadSuccessEvent;
import me.ruslanys.vkmusic.exception.DownloadException;
import me.ruslanys.vkmusic.property.DownloaderProperties;
import me.ruslanys.vkmusic.services.DownloadService;
import me.ruslanys.vkmusic.services.PropertyService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;

import java.io.File;
import java.util.List;
import java.util.concurrent.*;

/**
 * @author Ruslan Molchanov (ruslanys@gmail.com)
 */
@Slf4j

@Service
public class DefaultDownloadService implements DownloadService {

    private final ApplicationEventPublisher publisher;

    private final VkClient vkClient;
    private final PropertyService propertyService;


    @Autowired
    public DefaultDownloadService(@NonNull ApplicationEventPublisher publisher,
                                  @NonNull VkClient vkClient,
                                  @NonNull PropertyService propertyService) {
        this.publisher = publisher;
        this.vkClient = vkClient;
        this.propertyService = propertyService;
    }

    @Async
    @Override
    public void downloadAsync(List<Audio> audios) {
        download(audios);
    }

    @Override
    public void download(List<Audio> audios) {
        log.info("Download queue: {}", audios.size());
        vkClient.fetchUrls(audios);

        DownloaderProperties properties = propertyService.get(DownloaderProperties.class);
        download(properties, audios);
    }

    @SneakyThrows
    private void download(DownloaderProperties properties, List<Audio> audios) {
        log.info("Download pool-size: {}", properties.getPoolSize());
        log.info("Download destination: {}", properties.getDestination());

        File destinationFolder = new File(properties.getDestination());
        if (!destinationFolder.exists() && !destinationFolder.mkdirs()) {
            throw new IllegalStateException("Can not create destination folder.");
        }

        ExecutorService executor = Executors.newFixedThreadPool(properties.getPoolSize());
        CompletionService<DownloadResult> completionService = new ExecutorCompletionService<>(executor);

        for (Audio audio : audios) {
            completionService.submit(new DownloadTask(destinationFolder, audio));
        }

        int n = audios.size();
        for (int i = 0; i < n; i++) {
            try {
                DownloadResult result = completionService.take().get();
                publisher.publishEvent(new DownloadSuccessEvent(this, result));
            } catch (ExecutionException ex) {
                publisher.publishEvent(new DownloadFailEvent(this, (DownloadException) ex.getCause()));
            }
        }

        executor.shutdown();
        executor.awaitTermination(1, TimeUnit.MINUTES);

        publisher.publishEvent(new DownloadFinishEvent(this, audios));
    }


}

File: src/main/kotlin/me/ruslanys/vkmusic/service/DefaultDownloadService.kt
package me.ruslanys.vkmusic.service

import me.ruslanys.vkmusic.component.VkClient
import me.ruslanys.vkmusic.domain.Audio
import me.ruslanys.vkmusic.event.DownloadFailEvent
import me.ruslanys.vkmusic.event.DownloadInProgressEvent
import me.ruslanys.vkmusic.event.DownloadSuccessEvent
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.context.ApplicationEventPublisher
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service
import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileOutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.Executor

@Service
class DefaultDownloadService(
        private val vkClient: VkClient,
        private val publisher: ApplicationEventPublisher,
        @Qualifier("downloadExecutor") private val executor: Executor
) : DownloadService {

    @Async
    override fun download(destination: File, audio: Audio) {
        download(destination, listOf(audio))
    }

    @Async
    override fun download(destination: File, audioList: List<Audio>) {
        log.info("Download [{}]", audioList.joinToString { it.id.toString() })
        vkClient.fetchUrls(audioList)

        if (!destination.exists() || !destination.isDirectory) {
            throw IllegalArgumentException("Destination path is incorrect.")
        }

        for (audio in audioList) {
            executor.execute {
                downloadFile(destination, audio)
            }
        }
    }

    private fun downloadFile(destination: File, audio: Audio) {
        log.info("Download file {}", audio.url)
        publisher.publishEvent(DownloadInProgressEvent(this, audio))

        try {
            val connection = URL(audio.url).openConnection() as HttpURLConnection
            connection.connectTimeout = CONNECTION_TIMEOUT
            connection.readTimeout = CONNECTION_TIMEOUT

            val file = File(destination, audio.filename())
            BufferedInputStream(connection.inputStream, CONNECTION_BUFFER_SIZE).use { input ->
                BufferedOutputStream(FileOutputStream(file), CONNECTION_BUFFER_SIZE).use { output ->
                    val buff = ByteArray(CONNECTION_BUFFER_SIZE)
                    var len = 0
                    while (len != -1) {
                        len = input.read(buff)
                        if (len > 0) output.write(buff, 0, len)
                    }
                }
            }

            publisher.publishEvent(DownloadSuccessEvent(this, audio, file))
        } catch (e: Exception) {
            publisher.publishEvent(DownloadFailEvent(this, audio, e))
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(DefaultDownloadService::class.java)

        private const val CONNECTION_TIMEOUT = 10000
        private const val CONNECTION_BUFFER_SIZE = 5120
    }

}

fun Audio.filename(): String {
    val regex = "[!\"#$%&'()*+,\\-/:;<=>?@\\[\\]^_`{|}~]".toRegex()
    val formattedArtist = artist.trim().replace(regex, "")
    val formattedTitle = title.trim().replace(regex, "")

    return "${formattedArtist.take(10)} - ${formattedTitle.take(20)}.mp3"
}